### PR TITLE
refactor(v2): allow adding plugins depending on condition

### LINKED ISSUE #3663: Allow an ergonomic way to filter plugins between dev & prod.
## 🚀 Feature

Allow an ergonomic way to filter plugins between dev & prod. We were asked to open an issue about this in Discord - ref https://github.com/gulpjs/gulpjs.github.io/issues/123#issuecomment-718696325

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

## Motivation

For gulp's website, we want to have a series of plugins that are only applied at production-build time (analytics, gtag, sitemap, etc).

We are currenly doing it with this code, on an old 2-alpha version:

```
plugins: [
    isProd && ['@docusaurus/plugin-google-analytics', {}],
    isProd && ['@docusaurus/plugin-google-gtag', {}],
    isProd && ['@docusaurus/plugin-sitemap', {}],
]
```

This became invalid in a more recent alpha version that we are upgrading into because those resolve to `false` in development.

## Pitch

I think it is fairly common that people will want to have some plugins they don't want active during development (e.g. carbon-ads plugin) and they will want a concise way to filter those out.

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

Fixes #3663, provides an easy way to enable plugins based on external conditions (as was available in earlier versions of Docusaurus v2).

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

See #3663

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus/src/server/configValidation.ts [hunk 1]
```diff
@@ -40,6 +40,7 @@ export const DEFAULT_CONFIG: Pick<
 const PluginSchema = Joi.alternatives().try(
   Joi.string(),
   Joi.array().items(Joi.string().required(), Joi.object().required()).length(2),
+  Joi.bool(), // In case of conditional adding of plugins.
```
Comment (slorber) [inline packages/docusaurus/src/server/configValidation.ts:43]: not sure but I think it's possible to do `Joi.bool().equal(false)`. I think passing true should be rejected

Comment (lex111) [reply packages/docusaurus/src/server/configValidation.ts]: Agree, fixed it.

**Comment by lex111** [comment]
> This job has been blocked because resource-class:large is not available on your plan. Please upgrade to continue building.

What is it?

**Comment by lex111** [comment]
Ah, these CI errors occurs because I made a PR from my fork, because _for some reason_ I was excluded from collaborators of Docusaurus repo :cry:

**Pull request closed without merging**